### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by roynalnaruto**
Refactor the SBV:
- Use `tracing` instead of `log`
    - [x] TODO: replace logs with tracing API
- Add `VerificationError` for better error handling
    - [x] Make functions `verify` and `handle_block` fallible and appropriately handle errors
- Several minor refactors

### CODE REVIEW

File: src/bin/trace-verifier/commands/run_file.rs [hunk 1]
```diff
@@ -1,9 +1,57 @@
...
+use crate::utils;
+
+/// A wrapper around [`eth_types::l2_types::BlockTrace`] so that we can re-implement custom
+/// deserializer for it.
+struct WrappedBlockTrace(BlockTrace);
+
+/// In the JSON format of [`eth_types::l2_types::BlockTrace`] appears a "chainID" key.
+const JSON_KEY_CHAIN_ID: &str = "chainID";
+
+impl<'de> Deserialize<'de> for WrappedBlockTrace {
```
Comment (lightsing) [inline src/bin/trace-verifier/commands/run_file.rs:16]: Why this rather than the fallback method? This deserialization is not a hotspot and only for compatibility. The new trait impl is much hard to understand for others.

File: src/database.rs [hunk 2]
```diff
@@ -93,26 +98,26 @@ impl DatabaseRef for ReadOnlyDB {
 
     /// Get block hash by block number.
     fn block_hash_ref(&self, _: u64) -> Result<B256, Self::Error> {
-        unimplemented!("BLOCKHASH is disabled")
```
Comment (lightsing) [inline src/database.rs:96]: We are actually not implementing this method, so `unimplemented!` is better I think.

**Review by lightsing — COMMENTED** [review]
using tracing replacing log is better. I think we could make the logging as a feature, so we can completely turn it off when it's disabled (like for the sp1 env). (I know it has a level filter and the log won't emit if filter is set higher than the log level. I want to not waste any cycles on extra things.)

File: src/database.rs (hunk 2 above)
Comment (roynalnaruto) [reply src/database.rs:96]: If the plan is to implement it, then I will revert this change

File: src/bin/trace-verifier/commands/run_file.rs (hunk 1 above)
Comment (roynalnaruto) [reply src/bin/trace-verifier/commands/run_file.rs:16]: Eventually are we guaranteed to use the "correct" JSON format for a trace? And disallow the JSON-RPC result JSON? I mean, was this fallback temporary?

**Comment by roynalnaruto** [comment]
@lightsing 
> so we can completely turn it off when it's disabled (like for the sp1 env)

I completely agree

> I want to not waste any cycles on extra things
Yes, I was a bit unsure how the cycle count may increase if the tracing subscriber is initialised, but the log level is set higher. To be more certain, we can feature-gate tracing/logging for dev purposes.

File: Cargo.toml [hunk 3]
```diff
@@ -4,20 +4,23 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
-log = "0.4"
 hex = "0.4"
+itertools = "0.13.0" # for compatibility msrv 1.75
+#log = "0.4"
```
Comment (roynalnaruto) [inline Cargo.toml:9]: remove?

File: Cargo.toml [hunk 4]
```diff
@@ -63,18 +68,20 @@ bin-deps = [
...
     "tokio/macros",
-    "tokio/rt-multi-thread"
+    "tokio/rt-multi-thread",
+    "url"
 ]
 profiling = ["pprof"]
 debug-account = ["csv", "revm/serde"]
 debug-storage = ["csv", "revm/serde"]
+dev = ["tracing", "tracing-subscriber"]
+#dev = []
```
Comment (roynalnaruto) [inline Cargo.toml:84]: remove?

File: src/utils/ext.rs [hunk 5]
```diff
@@ -770,7 +775,7 @@ mod tests {
         let h160 = eth_types::H160::from(array::from_fn(|i| i as u8));
         let serialized = rkyv::to_bytes::<_, 20>(&h160).unwrap();
         let archived: &ArchivedH160 = unsafe { rkyv::archived_root::<H160>(&serialized[..]) };
-        assert_eq!(archived, &h160);
+        // assert_eq!(archived, &h160);
```
Comment (roynalnaruto) [inline src/utils/ext.rs:778]: mistakenly commented?

File: src/lib.rs [hunk 6]
```diff
@@ -1,17 +1,24 @@
+//! Stateless Block Verifier
+
 #![deny(missing_docs)]
 #![deny(missing_debug_implementations)]
-//! Stateless Block Verifier
+// #![allow(unused_imports)]
+// #![allow(unused_variables)]
```
Comment (roynalnaruto) [inline src/lib.rs:6]: remove?

File: src/executor/mod.rs [hunk 7]
```diff
@@ -69,21 +85,29 @@ impl EvmExecutor {
                     .build();
                 cycle_tracker_end!("build Evm");
 
-                trace!("handler cfg: {:?}", revm.handler.cfg);
+                dev_trace!("handler cfg: {:?}", revm.handler.cfg);
 
                 cycle_tracker_start!("transact_commit");
-                let result = revm.transact_commit().unwrap(); // TODO: handle error
+                #[allow(unused_variables)]
```
Comment (roynalnaruto) [inline src/executor/mod.rs:91]: why is this required?

File: src/bin/trace-verifier/utils.rs [hunk 8]
```diff
@@ -10,11 +13,11 @@ pub fn verify(
     fork_config: &HardforkConfig,
     disable_checks: bool,
     log_error: bool,
-) -> bool {
+) -> Result<bool, VerificationError> {
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/utils.rs:16]: now that we use a result, we can switch to:
```suggestion
) -> Result<(), VerificationError> {
```

File: src/bin/trace-verifier/utils.rs [hunk 9]
```diff
@@ -55,31 +58,34 @@ pub fn verify(
...
-        error!("Root mismatch");
+        dev_error!("Root after in trace: {:x}", root_after);
+        dev_error!("Root after in revm: {:x}", revm_root_after);
+        dev_error!("Root mismatch");
+
         if !log_error {
             std::process::exit(1);
         }
-        return false;
+        return Ok(false);
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/utils.rs:74]: return err instead of `Ok(false)`:
```suggestion
        return Err(VerificationError::StateRootMismatch(root_after, revm_root_after));
```

File: src/bin/trace-verifier/utils.rs [hunk 10]
```diff
@@ -55,31 +58,34 @@ pub fn verify(
...
         let elapsed = last_time.elapsed().as_secs_f64();
         let bps = blocks as f64 / elapsed;
-        warn!("Verifying avg speed: {:.2} bps", bps);
+
+        dev_warn!("Verifying avg speed: {:.2} bps", bps);
         *last_time = Instant::now();
     }
 
-    true
+    Ok(true)
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/utils.rs:90]:
```suggestion
    Ok(())
```

File: src/bin/trace-verifier/utils.rs
Comment (roynalnaruto) [inline src/bin/trace-verifier/utils.rs:1]: we can also add an error variant to propagate rkyv errors and remove those `unwrap`s

File: src/bin/trace-verifier/commands/run_rpc.rs [hunk 11]
```diff
@@ -90,15 +91,15 @@ impl RunRpcCommand {
...
+                        dev_info!(
                             "worker#{idx}: load trace for block #{block_number}({})",
                             l2_trace.header.hash.unwrap()
                         );
 
                         let success = tokio::task::spawn_blocking(move || {
                             utils::verify(l2_trace, &fork_config, disable_checks, is_log_error)
                         })
-                        .await?;
+                        .await??;
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/commands/run_rpc.rs:102]: once you change `verify` to be `Result<(), VerificationError>` then you can use `is_ok()` here, instead of `if !success`

**Review by roynalnaruto — COMMENTED** [review]
additional review round

File: src/executor/mod.rs (hunk 7 above)
Comment (i-m-aditya) [reply src/executor/mod.rs:91]: Getting rid of unused_variables warning when we don't run it with feature-dev.

File: src/utils/ext.rs (hunk 5 above)
Comment (i-m-aditya) [reply src/utils/ext.rs:778]: Linting issue, no issue with compilation.

File: src/bin/trace-verifier/commands/run_file.rs [hunk 12]
```diff
@@ -37,10 +37,13 @@ impl RunFileCommand {
...
 
             let fork_config = fork_config(trace.chain_id);
-            tokio::task::spawn_blocking(move || {
+            if let Err(err) = tokio::task::spawn_blocking(move || {
                 utils::verify(trace, &fork_config, disable_checks, false)
             })
-            .await??;
+            .await?
+            {
+                dev_error!("Verification failed: {:?}", err);
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/commands/run_file.rs:45]: The `dev_error` here is not necessary. The guest program in `sproll-evm` depends only on the library from `stateless_block_verifier`. This binary `trace-verifier` is just a helper for us to test SBV. So we can remove this.

Instead, similar to the `run_rpc.rs` you should do:
```rust
let mut guard = error_log.as_ref().unwrap().lock().await;
guard.write_all(/* */).await?;
```

File: src/bin/trace-verifier/commands/run_file.rs [hunk 13]
```diff
@@ -1,6 +1,6 @@
 use clap::Args;
 use eth_types::l2_types::BlockTrace;
-use stateless_block_verifier::{dev_info, HardforkConfig};
+use stateless_block_verifier::{dev_error, dev_info, HardforkConfig};
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/commands/run_file.rs:3]: we don't need the `dev_error` here

File: src/bin/trace-verifier/commands/run_rpc.rs [hunk 14]
```diff
@@ -96,16 +96,20 @@ impl RunRpcCommand {
...
+                        {
                             let mut guard = error_log.as_ref().unwrap().lock().await;
                             guard
                                 .write_all(format!("{block_number}\n").as_bytes())
                                 .await?;
+                            dev_error!(
+                                "Verification failed for block #{}: {:?}",
+                                block_number,
+                                err
+                            );
```
Comment (roynalnaruto) [inline src/bin/trace-verifier/commands/run_rpc.rs:112]: don't need `dev_error` here

**Review by roynalnaruto — COMMENTED** [review]
Another round

**Review by i-m-aditya — APPROVED** [review]
lgtm
